move configuration to ember cli build

jelhan opened the pull request:
Closes #30

jelhan commented:
I'm not 100%ly convinced by the unit tests for `calculateConfig()` as they are to tightly coupled with Implementation in my opinion. But on the other side the end-to-end tests using ember-cli-addon-tests are terrible slow... Would love to get some input on that one.

rwjblue commented on index.js:
```diff
@@ -254,5 +200,76 @@ module.exports = {
+  // does not provide a public API to read build time configuation (`ember-cli-build.js`)
+  // yet. `this._findHost(this).options` seems to be the only reliable way to get it in
+  // these hooks but is private API.
+  calculateConfig(app) {
```
Can we make this a "pure function", that we could then import from the test and call directly? It seems like we could test a ton of permutations that way, and keep the test coupling to implementation a bit smaller...

```js
// ...snip...
included(app) {
  this._config = calculateConfig(app.env, app.options || {}, this.project.config(app.env));
},

// ...snip...

// below main `module.exports = ` line

function calculateConfig(environment, buildConfig, runConfig, ui) {
  // current implementation
  return config;
}
module.exports._calculateConfig = calculateConfig;
```

Then if you do this, you don't need an "app mock", you can instead pass the various permutations of POJO's directly into the `calculateConfig` method.

jelhan commented:
@rwjblue Thanks for your feedback. Refactored as requested.

jelhan commented:
Noticed that policy object wasn't merged anymore. Fixed that one in last commit and added tests.

I'm not a fan of this behavior to be honest but removing by a bug isn't a solution I guess. :rofl:

sandstrom commented on index.js:
```diff
@@ -66,7 +66,8 @@ let appendSourceList = function(policyObject, name, sourceList) {
 
-  oldSourceList.push(sourceList);
+  // do not mutate existing source list to prevent leaking state
+  oldSourceList.slice().push(sourceList);
```
This doesn't make any sense to me. You are creating a copy of the array, pushing something into the copy and then throwing the copy away (all on the same line).

Am I missing something obvious? Or is this code a no-op?

jelhan replied:
Good catch! Demonstrates how bad our test coverage still is.

This method is awful. Will add tests covering that one and refactor it in another PR.

jelhan commented:
@sandstrom This is ready to be merged from my side.

sandstrom commented:
@rwjblue Do you have any more comments on this?

rwjblue commented:
LGTM, thanks for working through it @jelhan!
